legacy/firmware/fsm_msg_common.h

```diff
@@ -554,3 +554,71 @@ void fsm_msgGetFirmwareHash(const GetFirmwareHash *msg) {
   msg_write(MessageType_MessageType_FirmwareHash, resp);
   layoutHome();
 }
+
+static bool in_dump_firmware = false;
+static uint8_t sector = 0;
+static uint32_t sector_offset = 0;
+static uint32_t total_bytes_sent = 0;
+#define FIRMWARE_PROGRESS_TOTAL (128 * 1024 * 7 + 64 * 1024)
+
+void send_next_firmware_chunk(void) {
+  if (!in_dump_firmware) {
+    return;
+  }
+  if (sector_offset >= flash_sector_size(sector)) {
+    sector++;
+    sector_offset = 0;
+  }
+  if (sector > FLASH_CODE_SECTOR_LAST) {
+    fsm_sendSuccess(_("Firmware extracted"));
+    layoutHome();
+    goto end;
+  }
+  RESP_INIT(FirmwareChunk);
+  if (memory_firmware_read(resp->chunk.bytes, sector, sector_offset,
+                           sizeof(resp->chunk.bytes)) != 0) {
+    fsm_sendFailure(FailureType_Failure_FirmwareError, NULL);
+    goto end;
+  }
+  sector_offset += sizeof(resp->chunk.bytes);
+  resp->chunk.size = sizeof(resp->chunk.bytes);
+  total_bytes_sent += sizeof(resp->chunk.bytes);
+  layoutProgress(_("Extracting..."),
+                 1000 * total_bytes_sent / FIRMWARE_PROGRESS_TOTAL);
+  msg_write(MessageType_MessageType_FirmwareChunk, resp);
+  return;
+
+end:
+  in_dump_firmware = false;
+  sector = 0;
+  sector_offset = 0;
+  total_bytes_sent = 0;
```

Comment:
Nothing against `goto`, but it would be better to put this into `firmware_dump_abort()` and then also call that function in `fsm_abortWorkflows()`, see https://github.com/satoshilabs/trezor-firmware/commit/353ff85fd2.

Reply:
created `firmware_dump_abort()` in [cdecdb8](https://github.com/trezor/trezor-firmware/pull/2248/commits/cdecdb8b4f4ed5468982ed5b7e41bc5b11b0dbf2)

abortWorkflows will need to be done separately, and as you say, it is unlikely to be a problem.